src/map_io/map_resources_packet.cc

```diff
@@ -51,11 +52,14 @@ void MapResourcesPacket::read(FileSystem& fs, EditorGameBase& egbase) {
 
 			for (uint16_t y = 0; y < map->get_height(); ++y) {
 				for (uint16_t x = 0; x < map->get_width(); ++x) {
-					DescriptionIndex const id = fr.unsigned_8();
-					ResourceAmount const amount = fr.unsigned_8();
-					ResourceAmount const start_amount = fr.unsigned_8();
-					const auto fcoords = map->get_fcoords(Coords(x, y));
-					map->initialize_resources(fcoords, smap[id], start_amount);
+					const uint32_t id = (packet_version >= 2) ? fr.unsigned_32() : fr.unsigned_8();
+					const uint16_t amount = (packet_version >= 2) ? fr.unsigned_16() : fr.unsigned_8();
+					const uint16_t start_amount =
```

Comment:
If a higher value range is needed, the definitions of `DescriptionIndex` and `ResourceAmount` should be changed. Otherwise they might be truncated in other locations (e.g. in `initialize_resources`).
Also the smap index is only `uint8_t`, so don't think, that this works correctly.

Reply:
The problem still persists: `initialize_resources` and `set_resources` take a `ResourceAmount` which is defined as `uint8_t` in `logic/widelands.h` or `logic/field.h`. So using `uint32_t` in the packet won't prevent an overflow in the actual game data. 
I'm not sure, what you are trying to achieve by changing the types and which value actually overflows at which point.

Reply:
There are actually two problems here. First, the smap lookup uses the `[]` operator which creates an element if none was present yet, so if the loaded value is kNoResource then it creates a new mapping from kNoResource to 0, effectively replacing kNoResource with a random resource (typically coal). Second, DescriptionIndex used to be a uint8 and was changed to uint16 a while ago, but this packet persists it with a mixture of uint8 and uint16. So there's a risk that the uint16 value is truncated when stored as uint8. Most saveload functions use uint32 to persist DescriptionIndex.

There can be no overflow with ResourceAmount since the value that is saved by the packet is a uint8; so we're saving a uint8 as a uint32, and although we load a uint32 the actual value range is only uint8 at the moment – so no danger of overflow. If in the future we change ResourceAmount to uint16 or uint32 then the resource packet will still work correctly because it can now already accommodate the larger value range.

Reply:
Thanks for the explanation.
So the current problem is with DescriptionIndex and the other values are for forward compatibility.